sstables/sstable_set.cc

```diff
@@ -727,6 +727,67 @@ sstable_set_impl::create_single_key_sstable_reader(
     return make_combined_reader(schema, std::move(permit), std::move(readers), fwd, fwd_mr);
 }
 
+flat_mutation_reader
+time_series_sstable_set::create_single_key_sstable_reader(
+        column_family* cf,
+        schema_ptr schema,
+        reader_permit permit,
+        utils::estimated_histogram& sstable_histogram,
+        const dht::ring_position& pos,
+        const query::partition_slice& slice,
+        const io_priority_class& pc,
+        tracing::trace_state_ptr trace_state,
+        streamed_mutation::forwarding fwd_sm,
+        mutation_reader::forwarding fwd_mr) const {
+    // First check if the optimized algorithm for TWCS single partition queries can be applied.
+    // Multiple conditions must be satisfied:
+    // 1. The sstables must be sufficiently modern so they contain the min/max column metadata.
+    // 2. The schema cannot have static columns, since we're going to be opening new readers
+    //    into new sstables in the middle of the partition query. TWCS sstables will usually pass
+    //    this condition.
+    // 3. The sstables cannot have partition tombstones for the same reason as above.
+    //    TWCS sstables will usually pass this condition.
+    using sst_entry = std::pair<position_in_partition, shared_sstable>;
+    if (schema->has_static_columns()
+            || std::any_of(_sstables->begin(), _sstables->end(),
+                [] (const sst_entry& e) {
+                    return e.second->get_version() < sstable_version_types::md
```

Comment:
This condition shouldn't be exposed. Please follow up with setting sstable::_position_range to all rows for < md, so that it will naturally select all sstables.

See this comment:

```c++
    // Notes:
    // - we are going to rely on min/max column names for
    //   clustering filtering only from md-format sstables,
    //   see sstable::may_contain_rows().
    //   We choose not to clear_incorrect_min_max_column_names
    //   from older versions here as this disturbs sstable unit tests.
    //
    // - now that we store min/max metadata for range tombstones,
    //   their size may differ.
    if (!_schema->clustering_key_size()) {
        clear_incorrect_min_max_column_names();
        return;
    }
```

If we clear it here, the rest works naturally (but tests may need adjustment).

Reply:
AFAIU right now if the metadata is cleared, `sstable::_position_range` is set to `all_clustered_rows()`.
But in that case I don't want the new reader to be used at all - I want to go with the generic reader.
I could replace the version check with `has_correct_min_max_column_names()`:

```
    bool has_correct_min_max_column_names() const noexcept {
        return _version >= sstable_version_types::md;
    }
```